src/operations/blas2/gemv.hpp

```diff
@@ -80,6 +80,244 @@ SYCL_BLAS_INLINE void AddSetColumns<rhs_t>::adjust_access_displacement() {
+Gemv<lhs_t, matrix_t, vector_t, is_transposed, cache_line_size, work_per_thread>::eval(local_memory_t local_mem, cl::sycl::nd_item<1> ndItem) {
+
+  const index_t local_id = ndItem.get_local_id(0);
+  const index_t local_range = ndItem.get_local_range(0);
+  const index_t group_id = ndItem.get_group(0);
+  const index_t group_range = ndItem.get_group_range(0);
+  const index_t lda = matrix_.getSizeL();
+  const index_t nc_dim = is_transposed ? matrix_.get_size_col() : matrix_.get_size_row();
+  const index_t c_dim = is_transposed ? matrix_.get_size_row() : matrix_.get_size_col();
+
+  // ID of the group in the non-contracting dimension of the global grid of WGs
+  const index_t nc_group_id = group_id / WGs_per_C_;
+  // ID of the group in the contracting dimension in the global grid of WGs
+  const index_t c_group_id = group_id % WGs_per_C_;
+  
+  // Threads pre-fetch portions of X into local group-shared memory
+  const index_t x_vec_index = local_id + c_group_id * local_range;
+  local_mem[local_id] = x_vec_index < vector_.get_size() ? vector_.eval(x_vec_index) : 0;
+
+  // Barrier to ensure whole portion of vector X is in local memory
+  ndItem.barrier(cl::sycl::access::fence_space::local_space);
+
+  // Non-contracting dimension index
+  const index_t nc_dim_index = local_id + nc_group_id * local_range;
+
+  // Stores the sum for the partial dot product computation
+  value_t sum = 0;
+
+  // In the non transposed case
+  if (!is_transposed) {
+
+    // Make sure nc_dim_index is within bounds
+    // This closes threads that would be calculating beyond the final row
+    if (nc_dim_index >= nc_dim) {
+      return 0;
+    }
+
+    // Calculate the matrix index
+    index_t mat_index = nc_dim_index + c_group_id * local_range * lda;
+    
+    const index_t last_c_dim_id = std::min(c_dim - c_group_id * local_range, local_range);
+
+    // Computes the partial dot product for a row
+    for (index_t c_dim_id = 0; c_dim_id < last_c_dim_id; ++c_dim_id) {
+      sum += ProductOperator::eval(matrix_.template eval<true>(mat_index), local_mem[c_dim_id]);
+      mat_index += lda;
+    }
+
+    const index_t out_index = nc_dim_index + (c_group_id * nc_dim);
+    return lhs_.eval(out_index) = sum;
+  }
+  else { // In the transposed case 
+
+    constexpr int cache_capacity = cache_line_size / sizeof(value_t);
+
+    const int tile_c_loops = local_range / cache_capacity;
+
+    // Pointer to scratch space used for loading input matrix and transposing it on-the-fly
+    value_t* matrix_scratch = local_mem.localAcc.get_pointer() + local_range;
+
+    for (index_t c_tile_id = 0; c_tile_id < tile_c_loops; ++c_tile_id) {
+
+      // Extract a matrix block from global memory
+      extract_input_block(matrix_scratch, ndItem, c_tile_id);
+
+      // Ensure memory synchronization within work group
+      ndItem.barrier(cl::sycl::access::fence_space::local_space);
+
+      index_t mat_index = local_id;
```

Comment:
The declaration could be moved out, but the assignment cant, `mat_index` has to be reset on each iteration of the outer loop since it is in local memory, which gets updated on each iteration of the outer loop.

Reply:
Ah that is my mistake, I missed it being modified in the inner loop!